src/supervisor.tex

```diff
@@ -1487,6 +1519,98 @@ \subsection{Virtual Address Translation Process}
+implicit reads observe writes to the same memory locations, an SFENCE.VMA
+instruction must be executed after the writes to flush the relevant cached
+translations.
+
+The address-translation cache cannot be used in step 7; accessed and
+dirty bits may only be updated in memory directly.
+
+\begin{commentary}
+  The address-translation cache abstraction was used less explicitly in earlier
+  revisions of this specification, but it was made more explicit in order to
+  faciliate defining the behavior of NAPOT PTEs in Chapter~\ref{napotpages}.
+  In particular, an implicit load that hits on an address-translation cache
+  entry previously fetched due to a separate NAPOT PTE aliasing the 4KiB region
+  originally targeted is not easily described in terms of access ordering
+  alone.  In such a case, the PTE value being loaded was originally written to
+  an address that differs from the address of the PTE targeted by the
+  address-translation algorithm, and this complicates the memory ordering
+  analysis.  NAPOT PTEs notwithstanding, the inclusion of more explicit
+  language regarding the address-translation cache is not intended to imply any
+  actual change in behavior from prior versions of this specification.
+
+  It is permitted for multiple address-translation cache entries to co-exist
+  for the same address.  This represents the fact that in a conventional TLB
+  hierarchy, it is possible for multiple entries to match a single address if, for
+  example, a page is upgraded to a superpage without first clearing the
+  original non-leaf PTE's valid bit and executing an SFENCE.VMA with {\em
+  rs1}={\tt x0}, or if multiple TLBs exist in parallel at a given level of the
+  hierarchy.  In this case, just as if an SFENCE.VMA is not executed between
+  a write to the memory-management tables and subsequent implicit read of the
+  same address: it is unpredictable whether the old non-leaf PTE or the new leaf
+  PTE is used, but the behavior is otherwise well defined.
+\end{commentary}
+
+Implementations may also execute the address-translation algorithm
+speculatively at any time, for any virtual address, as long as {\tt satp}.MODE
+indicates that virtual addressing is enabled.
+Such speculative executions have the effect of pre-populating the
+address-translation cache.  Implementations must only perform implicit reads of
+the translation data structures pointed to by the current contents of the {\tt
+satp} register or a subsequent valid (V=1) translation data structure entry.
+
+Speculative executions of the address-translation algorithm behave as
+non-speculative executions of the algorithm do, except that they must not set the
+dirty bit for a PTE, they must not trigger an exception, and they must not create
+address-translation cache entries if those entries would have been invalidated
+by any SFENCE.VMA instruction executed by the hart since the speculative
+execution of the algorithm began.
```

Comment:
I'm not sure this description covers the case of speculation immediately after reset, unless we're saying that the ATC entries cannot be populated until an SFENCE.VMA of the appropriate type is executed. One could certainly interpret the statement "at any time since the most recent SFENCE.VMA that subsumes that address" in line 1524 above as implying ATC entries cannot be established until the first SFENCE.VMA is executed. But I doubt that's what was intended?

Reply:
Daniel and I have a related question outstanding with Andrew whihc potentially would cover this.  Namely, can hardware start using a new satp value right away if it is written in M-mode (and start caching the results), or must hardware wait until execution has switched to the privilege mode relevant to satp (i.e S/U modes)?  (For vsatp and hgatp the same question would be wrt M/HS modes being "ok" versus VS/VU modes being required before hardware can start doing tablewalks using these.)

The latter choice clearly implies that speculative use of satp can't start immediately after reset (and not until M-mode is exited).  The former choice would also limit speculative use of satp until it at least is written by software.

At the same time, more generally, satp is undefined after reset, so it would be bad in any case to allow use of that initial satp value.

Reply:
Resolved in latest commit: translation cannot start until the "effective privilege mode" is U or S (or hypervisor equivalent)